### ISSUE TITLE
[JENKINS-49581] Pipeline with syntax error hung forever

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/kubernetes-plugin#2005]
<p>Hi,</p>

<p>Take this example:</p>

<pre>
podTemplate(
    label: <span class="code-quote">'postgres'</span>,
    containers: [
containerTemplate(
    name: <span class="code-quote">'postgres'</span>,
    image: <span class="code-quote">'postgres/latest'</span>,
    privileged: <span class="code-keyword">false</span>,
    alwaysPullImage: <span class="code-keyword">true</span>,
)
    ],
    node(<span class="code-quote">'postgres'</span>){
container(<span class="code-quote">'postgres'</span>) {
    stage(<span class="code-quote">"Pull"</span>) {
checkout scm
    }
}
    }
)
</pre>
<p>In my opinion there is syntax error here, didn't find any reference to `node` field in `podTemplate`. However this will wait forever for `pod` in k8s cluster while it's never created.</p>

<p>After changing this into proper syntax, it works ofc.</p>

<p>And this is what I see from console logs:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/41481/41481_image-2018-02-15-16-44-13-193.png" style="border: 0px solid black" /></span></p>

<p>Seems like vary common info, while it's waiting for `POD` creation.</p>

<p> </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jozue_noon">jozue_noon</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49581" target="_blank">Pipeline with syntax error hung forever</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: kubernetes-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2021-01-19T14:06:14+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 5
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Hi,</p>

<p>Take this example:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">podTemplate(
    label: <span class="code-quote">'postgres'</span>,
    containers: [
        containerTemplate(
            name: <span class="code-quote">'postgres'</span>,
            image: <span class="code-quote">'postgres/latest'</span>,
            privileged: <span class="code-keyword">false</span>,
            alwaysPullImage: <span class="code-keyword">true</span>,
        )
    ],
    node(<span class="code-quote">'postgres'</span>){
        container(<span class="code-quote">'postgres'</span>) {
            stage(<span class="code-quote">"Pull"</span>) {
                checkout scm
            }
        }
    }
)
</pre>
</div></div>
<p>In my opinion there is syntax error here, didn't find any reference to `node` field in `podTemplate`. However this will wait forever for `pod` in k8s cluster while it's never created.</p>

<p>After changing this into proper syntax, it works ofc.</p>

<p>And this is what I see from console logs:</p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/41481/41481_image-2018-02-15-16-44-13-193.png" style="border: 0px solid black" /></span></p>

<p>Seems like vary common info, while it's waiting for `POD` creation.</p>

<p> </p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Jenkins 2.101</code></li></ul>
<details><summary><i>1 attachment</i></summary>

- [image-2018-02-15-16-44-13-193.png](https://issues.jenkins.io/secure/attachment/41481/image-2018-02-15-16-44-13-193.png)
  > ![image-2018-02-15-16-44-13-193.png](https://issues.jenkins.io/secure/attachment/41481/image-2018-02-15-16-44-13-193.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/csanchez.png" title="csanchez's avatar" /> csanchez:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49581?focusedId=328858&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-328858">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>can you provide jenkins logs?
<a href="https://github.com/jenkinsci/kubernetes-plugin/#debugging" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/#debugging</a></p>
</pre>
</details>

</ul>
</details>

<p>can you provide jenkins logs?<br/>
<a href="https://github.com/jenkinsci/kubernetes-plugin/#debugging" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/kubernetes-plugin/#debugging</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jozue_noon:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49581?focusedId=328877&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-328877">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I'm afraid that there is no logs at all. Other piplines generate a lot of output to  this logger `org.csanchez.jenkins.plugins.kubernetes` however not UUT one. In "main" logger I get notification only after I abort job otherwise it's no mentioned, but I think it's not relevant, or ?</p>
</pre>
</details>

</ul>
</details>

<p>I'm afraid that there is no logs at all. Other piplines generate a lot of output to  this logger `org.csanchez.jenkins.plugins.kubernetes` however not UUT one. In "main" logger I get notification only after I abort job otherwise it's no mentioned, but I think it's not relevant, or ?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/csanchez.png" title="csanchez's avatar" /> csanchez:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49581?focusedId=328885&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-328885">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>ok, so it's a pipeline issue, not k8s plugin</p>
</pre>
</details>

</ul>
</details>

<p>ok, so it's a pipeline issue, not k8s plugin</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49581?focusedId=329454&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-329454">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Actually, this'd be in <tt>kubernetes-plugin</tt> - something about <tt>podTemplate</tt> must be executing that <tt>node</tt> block.</p>
</pre>
</details>

</ul>
</details>

<p>Actually, this'd be in <tt>kubernetes-plugin</tt> - something about <tt>podTemplate</tt> must be executing that <tt>node</tt> block.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>ecejjar:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49581?focusedId=348957&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-348957">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>What I think is happening is the following. Given the lazy evaluation of actual arguments intrinsic to Groovy, the groovy interpreter first tries to obtain all the actual arguments before calling the function. As part of that process it runs the node() step. Since podTemplate(label: 'postgres', ...) hasn't been called yet, there is no node with that label, and the node() step blocks indefinitely trying to get hold of such a node.</p>

<p>Notice the interpreter can't detect the wrong argument type in the call to podTemplate() since it first needs to evaluate the node() step. This kind of issues are catched by the compiler in a type-safe language.</p>

<p>If the node() step gave up and returned, the interpreter would catch the wrong argument type and the call to podTemplate() would fail.</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>What I think is happening is the following. Given the lazy evaluation of actual arguments intrinsic to Groovy, the groovy interpreter first tries to obtain all the actual arguments before calling the function. As part of that process it runs the node() step. Since podTemplate(label: 'postgres', ...) hasn't been called yet, there is no node with that label, and the node() step blocks indefinitely trying to get hold of such a node.</p>

<p>Notice the interpreter can't detect the wrong argument type in the call to podTemplate() since it first needs to evaluate the node() step. This kind of issues are catched by the compiler in a type-safe language.</p>

<p>If the node() step gave up and returned, the interpreter would catch the wrong argument type and the call to podTemplate() would fail.</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/vlatombe.png" title="vlatombe's avatar" /> vlatombe:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-49581?focusedId=403687&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-403687">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Can't be fixed for the reason explained in the comment above.</p>
</pre>
</details>

</ul>
</details>

<p>Can't be fixed for the reason explained in the comment above.</p>
